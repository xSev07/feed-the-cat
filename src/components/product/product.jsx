import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import {calculateBonus, declOfNum, deleteExtraSpaces} from '../../utils/common/common';
import {Declination, Multiplicity} from '../../const';
import withActiveStatus from "../../hocs/with-active-status/with-active-status.jsx";
import withMouseLeaveEvent from "../../hocs/with-mouse-leave-event/with-mouse-leave-event.jsx";

class Product extends PureComponent {
  constructor(props) {
    super(props);

    this._ended = props.product.balance <= 0;

    this._handlerProductClick = this._handlerProductClick.bind(this);
    this._calculateAdditionalCaption = this._calculateAdditionalCaption.bind(this);
    this._changeDisplayingQuestion = this._changeDisplayingQuestion.bind(this);
  }

  _handlerProductClick(evt) {
    evt.preventDefault();

    if (!this._ended) {
      this.props.onElementClick();
    }
  }

  _changeDisplayingQuestion(evt) {
    if (this.props.isActive) {
      this.props.onMouseLeave(evt);
    }
  }

  _calculateAdditionalCaption() {
    const {taste, description} = this.props.product;
    if (this._ended) {
      return `Печалька, ${taste} закончился`;
    }

    if (this.props.isActive) {
      return description;
    }

    return (
      <>
        Чего сидишь? Порадуй котэ,
        <a className='catalog__buy' href='#' onClick={this._handlerProductClick}> купи</a>
      </>
    );
  }

  render() {
    const {product, isActive, isMouseLeave} = this.props;
    const {id, caption, brand, taste, numberOfServings,
      weight, unit} = product;
    const captionMain = isMouseLeave ? `Котэ не одобряет?` : caption;

    const declServing = declOfNum(numberOfServings, Declination.SERVING);
    const mouseCount = calculateBonus(numberOfServings, Multiplicity.SERVING);
    const declMouse = declOfNum(mouseCount, Declination.MOUSE);

    return (
      <li className={deleteExtraSpaces(`catalog__item
        ${this._ended ? ` catalog__item--disabled` : ``}
        ${isActive ? ` catalog__item--selected` : ``}
      `)}
      >
        <input id={id} readOnly className='hidden' type='checkbox' checked={isActive}/>
        <a
          className='catalog__content'
          href='#'
          onClick={this._handlerProductClick}
          onMouseLeave={this._changeDisplayingQuestion}
          onMouseEnter={this._changeDisplayingQuestion}
        >
          <p className={`catalog__caption-main${isMouseLeave ? ` catalog__caption-main--showed-question` : ``}`}>
            {captionMain}
          </p>
          <h3 className='catalog__name-container'>
            <span className='catalog__name-brand'>{brand}</span>
            <br/>
            <span className='catalog__name-taste'>{taste}</span>
          </h3>
          <p className='catalog__spec'>
            <span className='catalog__quantity'>{numberOfServings} </span>
            {declServing}
            <br/>
            {mouseCount > 1 && <span className='catalog__quantity'>{mouseCount} </span>}
            {declMouse} в подарок
            <br/>
            {numberOfServings >= Multiplicity.SATISFIED_CUSTOMER && `заказчик доволен`}
          </p>
          <div className='catalog__picture-container'>
            <picture>
              <source type='image/webp' srcSet='assets/img/cat_preview.webp'/>
              <img className='catalog__picture' src='assets/img/cat_preview.png' width='260' height='275'
                alt={`${brand} ${taste}`}/>
            </picture>
          </div>
          <div className='catalog__size-container'>
            <p className='catalog__size-value'>{weight.toString().replace(`.`, `,`)}</p>
            <p className='catalog__size-unit'>{unit}</p>
          </div>
        </a>
        <p className='catalog__caption-additional'>
          {this._calculateAdditionalCaption()}
        </p>
      </li>
    );
  }
}

Product.propTypes = {
  product: PropTypes.shape({
    id: PropTypes.string.isRequired,
    caption: PropTypes.string.isRequired,
    brand: PropTypes.string.isRequired,
    taste: PropTypes.string.isRequired,
    numberOfServings: PropTypes.number.isRequired,
    weight: PropTypes.number.isRequired,
    unit: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    balance: PropTypes.number.isRequired,
  }),
  isActive: PropTypes.bool.isRequired,
  isMouseLeave: PropTypes.bool.isRequired,
  onElementClick: PropTypes.func.isRequired,
  onMouseLeave: PropTypes.func.isRequired,
};

export {Product};
export default withActiveStatus(withMouseLeaveEvent(Product));
