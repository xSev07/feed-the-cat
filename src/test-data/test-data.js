const products = [
  {
    id: `fg_small`,
    caption: `Сказочное заморское яство`,
    brand: `Нямушка`,
    taste: `c фуа-гра`,
    numberOfServings: 10,
    weight: 0.5,
    unit: `кг`,
    description: `Печень перекормленного гуся`,
    balance: 10,
  },
  {
    id: `fish_medium`,
    caption: `Сказочное заморское яство`,
    brand: `Нямушка`,
    taste: `c рыбой`,
    numberOfServings: 40,
    weight: 2,
    unit: `кг`,
    description: `Головы щучьи с чесноком да свежайшая сёмгушка.`,
    balance: 20,
  },
  {
    id: `chicken_large`,
    caption: `Сказочное заморское яство`,
    brand: `Нямушка`,
    taste: `c курой`,
    numberOfServings: 100,
    weight: 5,
    unit: `кг`,
    description: `Сочные ножки Буша`,
    balance: 0,
  },
];

export {products};
