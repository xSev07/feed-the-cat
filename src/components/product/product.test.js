import React from "react";
import renderer from "react-test-renderer";
import {Product} from "./product";
import {products} from "../../test-data/test-data";


describe(`Should Product render correctly`, () => {
  it(`Should render default card`, () => {
    const tree = renderer
      .create(
          <Product
            product={products[0]}
            isActive={false}
            isMouseLeave={false}
            onElementClick={() => {}}
            onMouseLeave={() => {}}
          />
      ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`Should render active card`, () => {
    const tree = renderer
      .create(
          <Product
            product={products[1]}
            isActive={true}
            isMouseLeave={false}
            onElementClick={() => {}}
            onMouseLeave={() => {}}
          />
      ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`Should render disabled card`, () => {
    const tree = renderer
      .create(
          <Product
            product={products[2]}
            isActive={false}
            isMouseLeave={false}
            onElementClick={() => {}}
            onMouseLeave={() => {}}
          />
      ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`Should render card with question`, () => {
    const tree = renderer
      .create(
          <Product
            product={products[1]}
            isActive={true}
            isMouseLeave={true}
            onElementClick={() => {}}
            onMouseLeave={() => {}}
          />
      ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
