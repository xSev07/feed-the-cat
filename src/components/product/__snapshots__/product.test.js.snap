// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should Product render correctly Should render active card 1`] = `
<li
  className="catalog__item catalog__item--selected "
>
  <input
    checked={true}
    className="hidden"
    id="fish_medium"
    readOnly={true}
    type="checkbox"
  />
  <a
    className="catalog__content"
    href="#"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <p
      className="catalog__caption-main "
    >
      Сказочное заморское яство
    </p>
    <h3
      className="catalog__name-container"
    >
      <span
        className="catalog__name-brand"
      >
        Нямушка
      </span>
      <br />
      <span
        className="catalog__name-taste"
      >
        c рыбой
      </span>
    </h3>
    <p
      className="catalog__spec"
    >
      <span
        className="catalog__quantity"
      >
        40
         
      </span>
      порций
      <br />
      <span
        className="catalog__quantity"
      >
        2
         
      </span>
      мыши
       в подарок
      <br />
    </p>
    <div
      className="catalog__picture-container"
    >
      <picture>
        <source
          srcSet="assets/img/cat_preview.webp"
          type="image/webp"
        />
        <img
          alt="Нямушка c рыбой"
          className="catalog__picture"
          height="275"
          src="assets/img/cat_preview.png"
          width="260"
        />
      </picture>
    </div>
    <div
      className="catalog__size-container"
    >
      <p
        className="catalog__size-value"
      >
        2
      </p>
      <p
        className="catalog__size-unit"
      >
        кг
      </p>
    </div>
  </a>
  <p
    className="catalog__caption-additional"
  >
    Головы щучьи с чесноком да свежайшая сёмгушка.
  </p>
</li>
`;

exports[`Should Product render correctly Should render card with question 1`] = `
<li
  className="catalog__item catalog__item--selected "
>
  <input
    checked={true}
    className="hidden"
    id="fish_medium"
    readOnly={true}
    type="checkbox"
  />
  <a
    className="catalog__content"
    href="#"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <p
      className="catalog__caption-main catalog__caption-main--showed-question"
    >
      Котэ не одобряет?
    </p>
    <h3
      className="catalog__name-container"
    >
      <span
        className="catalog__name-brand"
      >
        Нямушка
      </span>
      <br />
      <span
        className="catalog__name-taste"
      >
        c рыбой
      </span>
    </h3>
    <p
      className="catalog__spec"
    >
      <span
        className="catalog__quantity"
      >
        40
         
      </span>
      порций
      <br />
      <span
        className="catalog__quantity"
      >
        2
         
      </span>
      мыши
       в подарок
      <br />
    </p>
    <div
      className="catalog__picture-container"
    >
      <picture>
        <source
          srcSet="assets/img/cat_preview.webp"
          type="image/webp"
        />
        <img
          alt="Нямушка c рыбой"
          className="catalog__picture"
          height="275"
          src="assets/img/cat_preview.png"
          width="260"
        />
      </picture>
    </div>
    <div
      className="catalog__size-container"
    >
      <p
        className="catalog__size-value"
      >
        2
      </p>
      <p
        className="catalog__size-unit"
      >
        кг
      </p>
    </div>
  </a>
  <p
    className="catalog__caption-additional"
  >
    Головы щучьи с чесноком да свежайшая сёмгушка.
  </p>
</li>
`;

exports[`Should Product render correctly Should render default card 1`] = `
<li
  className="catalog__item "
>
  <input
    checked={false}
    className="hidden"
    id="fg_small"
    readOnly={true}
    type="checkbox"
  />
  <a
    className="catalog__content"
    href="#"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <p
      className="catalog__caption-main "
    >
      Сказочное заморское яство
    </p>
    <h3
      className="catalog__name-container"
    >
      <span
        className="catalog__name-brand"
      >
        Нямушка
      </span>
      <br />
      <span
        className="catalog__name-taste"
      >
        c фуа-гра
      </span>
    </h3>
    <p
      className="catalog__spec"
    >
      <span
        className="catalog__quantity"
      >
        10
         
      </span>
      порций
      <br />
      мышь
       в подарок
      <br />
    </p>
    <div
      className="catalog__picture-container"
    >
      <picture>
        <source
          srcSet="assets/img/cat_preview.webp"
          type="image/webp"
        />
        <img
          alt="Нямушка c фуа-гра"
          className="catalog__picture"
          height="275"
          src="assets/img/cat_preview.png"
          width="260"
        />
      </picture>
    </div>
    <div
      className="catalog__size-container"
    >
      <p
        className="catalog__size-value"
      >
        0,5
      </p>
      <p
        className="catalog__size-unit"
      >
        кг
      </p>
    </div>
  </a>
  <p
    className="catalog__caption-additional"
  >
    Чего сидишь? Порадуй котэ,
    <a
      className="catalog__buy"
      href="#"
      onClick={[Function]}
    >
       купи
    </a>
  </p>
</li>
`;

exports[`Should Product render correctly Should render disabled card 1`] = `
<li
  className="catalog__item catalog__item--disabled "
>
  <input
    checked={false}
    className="hidden"
    id="chicken_large"
    readOnly={true}
    type="checkbox"
  />
  <a
    className="catalog__content"
    href="#"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <p
      className="catalog__caption-main "
    >
      Сказочное заморское яство
    </p>
    <h3
      className="catalog__name-container"
    >
      <span
        className="catalog__name-brand"
      >
        Нямушка
      </span>
      <br />
      <span
        className="catalog__name-taste"
      >
        c курой
      </span>
    </h3>
    <p
      className="catalog__spec"
    >
      <span
        className="catalog__quantity"
      >
        100
         
      </span>
      порций
      <br />
      <span
        className="catalog__quantity"
      >
        5
         
      </span>
      мышей
       в подарок
      <br />
      заказчик доволен
    </p>
    <div
      className="catalog__picture-container"
    >
      <picture>
        <source
          srcSet="assets/img/cat_preview.webp"
          type="image/webp"
        />
        <img
          alt="Нямушка c курой"
          className="catalog__picture"
          height="275"
          src="assets/img/cat_preview.png"
          width="260"
        />
      </picture>
    </div>
    <div
      className="catalog__size-container"
    >
      <p
        className="catalog__size-value"
      >
        5
      </p>
      <p
        className="catalog__size-unit"
      >
        кг
      </p>
    </div>
  </a>
  <p
    className="catalog__caption-additional"
  >
    Печалька, c курой закончился
  </p>
</li>
`;
